### PR TITLE
KEYCLOAK-2606: document the different cordova modes of the js-adapter

### PR DESCRIPTION
**Pull request by gtudan**
Here's a first shot at the documentation for https://github.com/keycloak/keycloak/pull/5256.
I referenced the universal-links-plugin on how to setup the links, as I think their documentation is quiet good. 

Please let me know which parts could use some more details. Should we dig deeper into how to setup the apple-app-site-association/asset-links part?

### CODE REVIEW

**Comment by matthewhelmke** [comment]
I am going to allow someone from the engineering team, perhaps @stianst, to review first to make sure the content is complete and accurate. I will review this from a technical writing perspective after that.

**Review by stianst — DISMISSED** [review]

**Comment by stianst** [comment]
@matthewhelmke looks good from a technical perspective. Can you review please?

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: Keycloak supports hybrid mobile apps...

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: The default is `cordova`, which the adapter will automatically select if no adapter type has been configured and `window.cordova` is present. When logging in, it will open an https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-inappbrowser/[InApp Browser] that lets the user interact with {project_name} and afterwards returns to the app by redirecting to http://localhost. Because of this, you must whitelist this URL as a valid redirect-uri in the client configuration section of the Administration Console.

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: While this mode is easy to setup, it also has some disadvantages:
* The InApp-Browser is a browser embedded in the app and is not the phone's default browser. Therefore it will have different settings and stored credentials will not be available.
* The InApp-Browser might also be slower, especially when rendering more complex themes.
* There are security concerns to consider, before using this mode, such as that it is possible for the app to gain access to the credentials of the user, as it has full control of the browser rendering the login page, so do not allow its use in apps you do not trust.

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: Use this example app to help you get started: ...

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: The alternative mode `cordova-native` takes a different approach. It opens the login page using the system's browser. After the user has authenticated, the browser redirects back into the app using a special URL. From there, the {project_name} adapter can finish the login by reading the code or token from the URL.

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: _I am making an assumption here, that this applies to both modes of the Javascript adapter. Please correct if I am wrong._

This Javascript adapter requires two additional plugins, regardless of which mode you are using:

* https://github.com/google/cordova-plugin-browsertab[cordova-plugin-browsertab]: allows the app to open webpages in the system's browser
* https://github.com/e-imaxina/cordova-plugin-deeplinks[cordova-plugin-deeplinks]: allows the browser to redirect back to your app using special URLs

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: s/reliable/reliably

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: s/reliable/reliably

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (matthewhelmke) [inline securing_apps/topics/oidc/javascript-adapter.adoc]: * "cordova-native" - the library tries to open the login and registration page using the phone's system browser using the BrowserTabs cordova plugin. This requires extra setup for redirecting back to the app (see <<hybrid-apps-with-cordova>>) .

**Review by matthewhelmke — CHANGES_REQUESTED** [review]
Thank you for the information and taking the time to write it up! I have left several comments with edits for clarity and grammar.

File: securing_apps/topics/oidc/javascript-adapter.adoc
Comment (gtudan) [reply securing_apps/topics/oidc/javascript-adapter.adoc]: No, these plugins are only necessary if you're using the native mode.

**Comment by gtudan** [comment]
Hi Matthew,
thanks for checking the docs! If added the changes you suggested, could you take a look again?

**Comment by gtudan** [comment]
I could use some help on how to get the tests to pass. These are the problematic links:
```
checkExternalLinks(org.keycloak.documentation.test.SecuringAppsTest)  Time elapsed: 22.776 sec  <<< FAILURE!
* https://github.com/e-imaxina/cordova-plugin-deeplinks/blob/master/README.md#android-web-integration (invalid anchor android-web-integration)
* https://github.com/e-imaxina/cordova-plugin-deeplinks/blob/master/README.md#ios-web-integration (invalid anchor ios-web-integration)
```

I could remove the anchor links, but the document is quiet long and I feel they really help.

**Review by matthewhelmke — DISMISSED** [review]
Your changes look good. Thank you, and thanks again for writing this up and submitting it! The contribution is much appreciated.

I am asking around for help with the test problem. The links look fine to me, so hopefully someone else from our team can figure out what is going on.

**Comment by iankko** [comment]
@gtudan What happens if you escape the hash characters? IOW use something like this:
```
https://github.com/e-imaxina/cordova-plugin-deeplinks/blob/master/README.md\#android-web-integration
```
Didn't actually try it, but it could help to achieve, what you are searching for..

HTH, Jan

P.S.: Though escaping the Markdown anchor not to be recognized by asciidoc itself should help, per [this issue](https://github.com/asciidoctor/asciidoctor/issues/901) it seems, it isn't so unambiguous  / easy. That's why it needs trying first if it works.

**Comment by gtudan** [comment]
No, that didn't help. I guess the problem is that github uses javascript for their internal linking.
```html
<a id="user-content-android-web-integration" class="anchor" aria-hidden="true" href="#android-web-integration">
```
The anchors don't have a  `name` attribute and linking to the id doesn't work either.

So I guess the pragmatic options are:
* just linking the document (without anchor)
* disable checking of anchor links
* add these two links to the ignore list

**Comment by matthewhelmke** [comment]
The person who really knows our build pipeline, @stianst, is on holiday until 16 July and he is the one that would be able to modify link checking and ignore lists. 

I think that leaves two options: we wait until he gets back or we link the documents without the anchors. 

My instinct is to say that we remove the anchors and modify the paragraph, which will be more easily maintained in the future since we do not control the linked page and cannot predict when/if the anchors will be changed or removed. This is what I suggest:

`Please refer to the Android and iOS sections of the link:https://github.com/e-imaxina/cordova-plugin-deeplinks/blob/master/README.md[deeplinks plugin documentation] for further instructions.`

**Comment by gtudan** [comment]
Finally got the build to pass. Ready for merge once the feature hits master

**Comment by matthewhelmke** [comment]
@gtudan Excellent! Thank you for your persistence. This looks good.

When the feature hits master, if we don't merge this while we merge the feature (that is the goal), feel free to comment here and tag me or @stianst and we will merge right away.

**Review by matthewhelmke — APPROVED** [review]

**Review by stianst — APPROVED** [review]

**Pull request merged**
